// spinjs/lib/plugins/AngularPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var path = require("path");
var requireModule_1 = require("../requireModule");
var JSRuleFinder_1 = require("./shared/JSRuleFinder");
var AngularPlugin = (function () {
    function AngularPlugin() {
    }
    AngularPlugin.prototype.configure = function (builder, spin) {
        var stack = builder.stack;
        if (stack.hasAll(['angular', 'webpack'])) {
            var webpack = requireModule_1.default('webpack');
            var jsRuleFinder = new JSRuleFinder_1.default(builder);
            var tsRule = jsRuleFinder.findAndCreateTSRule();
            builder.config = spin.merge(builder.config, {
                module: {
                    rules: [
                        {
                            test: tsRule.test,
                            use: requireModule_1.default.resolve('angular2-template-loader')
                        }
                    ]
                },
                plugins: [
                    new webpack.ContextReplacementPlugin(/angular[\\\/]core[\\\/]@angular/, path.resolve('src'), {})
                ]
            });
            if (!stack.hasAny('dll') && stack.hasAny('web')) {
                builder.config = spin.merge({
                    entry: {
                        index: [require.resolve('./angular/angular-polyfill.js')]
                    }
                }, builder.config);
                var CheckerPlugin = requireModule_1.default('awesome-typescript-loader').CheckerPlugin;
                builder.config = spin.merge(builder.config, {
                    module: {
                        rules: [
                            {
                                test: /\.html$/,
                                loader: requireModule_1.default.resolve('html-loader')
                            }
                        ]
                    },
                    plugins: [new CheckerPlugin()]
                });
            }
        }
    };
    return AngularPlugin;
}());
exports.default = AngularPlugin;
