// spinjs/lib/plugins/ApolloPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var path = require("path");
var requireModule_1 = require("../requireModule");
var JSRuleFinder_1 = require("./shared/JSRuleFinder");
var persistPlugins;
var ApolloPlugin = (function () {
    function ApolloPlugin() {
    }
    ApolloPlugin.prototype.configure = function (builder, spin) {
        if (!builder.stack.hasAny('dll') && builder.stack.hasAll(['apollo', 'webpack'])) {
            var persistGraphQL = spin.options.persistGraphQL && !spin.test;
            if (builder.stack.hasAny(['server', 'web'])) {
                if (!persistPlugins) {
                    var PersistGraphQLPlugin = requireModule_1.default('persistgraphql-webpack-plugin');
                    var moduleName = path.resolve('node_modules/persisted_queries.json');
                    if (persistGraphQL) {
                        var clientPersistPlugin = new PersistGraphQLPlugin({
                            moduleName: moduleName,
                            filename: 'extracted_queries.json',
                            addTypename: true
                        });
                        var serverPersistPlugin = new PersistGraphQLPlugin({
                            moduleName: moduleName,
                            provider: clientPersistPlugin
                        });
                        persistPlugins = { client: clientPersistPlugin, server: serverPersistPlugin };
                    }
                    else {
                        var clientPersistPlugin = new PersistGraphQLPlugin({ moduleName: moduleName });
                        var serverPersistPlugin = new PersistGraphQLPlugin({ moduleName: moduleName });
                        persistPlugins = { client: clientPersistPlugin, server: serverPersistPlugin };
                    }
                }
            }
            builder.config = spin.merge(builder.config, {
                module: {
                    rules: [
                        {
                            test: /\.graphqls/,
                            use: requireModule_1.default.resolve('raw-loader')
                        },
                        {
                            test: /\.(graphql|gql)$/,
                            exclude: /node_modules/,
                            use: [requireModule_1.default.resolve('graphql-tag/loader')].concat(persistGraphQL ? [requireModule_1.default.resolve('persistgraphql-webpack-plugin/graphql-loader')] : [])
                        }
                    ]
                }
            });
            if (builder.stack.hasAny(['server', 'web'])) {
                var webpack = requireModule_1.default('webpack');
                if (persistGraphQL) {
                    var jsRuleFinder = new JSRuleFinder_1.default(builder);
                    var jsRule = jsRuleFinder.findAndCreateJSRule();
                    jsRule.use = spin.merge(jsRule.use, [requireModule_1.default.resolve('persistgraphql-webpack-plugin/js-loader')]);
                }
                builder.config = spin.merge(builder.config, {
                    plugins: [
                        new webpack.DefinePlugin({ __PERSIST_GQL__: persistGraphQL }),
                        builder.stack.hasAny('server') ? persistPlugins.server : persistPlugins.client
                    ]
                });
            }
        }
    };
    return ApolloPlugin;
}());
exports.default = ApolloPlugin;
