// spinjs/lib/plugins/WebAssetsPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var requireModule_1 = require("../requireModule");
var WebAssetsPlugin = (function () {
    function WebAssetsPlugin() {
    }
    WebAssetsPlugin.prototype.configure = function (builder, spin) {
        var stack = builder.stack;
        if (!stack.hasAny('dll') &&
            (stack.hasAll(['webpack', 'web']) || (stack.hasAll(['webpack', 'server']) && spin.options.ssr))) {
            builder.config = spin.merge(builder.config, {
                module: {
                    rules: [
                        {
                            test: /\.(png|ico|jpg|xml)$/,
                            use: {
                                loader: requireModule_1.default.resolve('url-loader'),
                                options: {
                                    name: '[hash].[ext]',
                                    limit: 100000
                                }
                            }
                        },
                        {
                            test: /\.woff(2)?(\?v=[0-9]\.[0-9]\.[0-9])?$/,
                            use: {
                                loader: requireModule_1.default.resolve('url-loader'),
                                options: {
                                    name: './assets/[hash].[ext]',
                                    limit: 100000
                                }
                            }
                        },
                        {
                            test: /\.(ttf|eot|svg)(\?v=[0-9]\.[0-9]\.[0-9])?$/,
                            use: {
                                loader: requireModule_1.default.resolve('file-loader'),
                                options: {
                                    name: './assets/[hash].[ext]'
                                }
                            }
                        }
                    ]
                }
            });
        }
        else if (!stack.hasAny('dll') && stack.hasAll(['webpack', 'server']) && !spin.options.ssr) {
            var ignoreLoader = requireModule_1.default.resolve('ignore-loader');
            builder.config = spin.merge(builder.config, {
                module: {
                    rules: [
                        {
                            test: /\.(png|ico|jpg|xml)$/,
                            use: {
                                loader: ignoreLoader
                            }
                        },
                        {
                            test: /\.woff(2)?(\?v=[0-9]\.[0-9]\.[0-9])?$/,
                            use: {
                                loader: ignoreLoader
                            }
                        },
                        {
                            test: /\.(ttf|eot|svg)(\?v=[0-9]\.[0-9]\.[0-9])?$/,
                            use: {
                                loader: ignoreLoader
                            }
                        }
                    ]
                }
            });
        }
    };
    return WebAssetsPlugin;
}());
exports.default = WebAssetsPlugin;
